Building a custom rviz display

DLu opened the pull request:
I hate RST.

DLu committed:
Building a custom rviz display

DLu commented:
Review please?

DLu commented:
@clalancette ? @audrow ? Any movement?

clalancette commented:
Sorry I never looked back at this.

I think this would be a good tutorial to have, but I also think it belongs in a series with other RViz changes.  I'm going to hold off on looking at this one until we land #4040 , at which point this will have a natural place to slot into.

clalancette commented:
OK, we've now landed #4040.  So I'm going to suggest that this gets moved into a tutorial under the new Tutorials/Intermediate/RViz directory.

DLu committed:
Move to RViz folder

DLu commented:
Rebased, and moved to the new `RViz` folder, and I even changed things up to follow the same directory structure as the existing tutorial in that folder.

ahcorde commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,419 @@
+Building a Custom RViz Display
+==============================
+
+`rviz_plugin_tutorial <https://github.com/MetroRobots/rviz_plugin_tutorial>`__: A tutorial on how to make rviz plugins
```
```suggestion
`rviz_plugin_tutorial <https://github.com/MetroRobots/rviz_plugin_tutorial>`__: A tutorial on how to make RViz plugins
```

ahcorde commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,419 @@
+Point2D Message
+---------------
+
+We'll be playing with a toy message defined in the ``rviz_plugin_tutorial_msgs``  package: ``Point2D.msg``:
```
```suggestion
We'll be playing with a toy message defined in the ``rviz_plugin_tutorial_msgs`` package: ``Point2D.msg``:
```

ahcorde commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,419 @@
+.. code-block:: c++
+
+
+   #pragma once
```
do you mind to use include guards? as we are doing on RViz ?

If you are using the common linters probaby it will require this

```cpp
#ifndef BLAH_H
#define BLAH_H

#endif  // BLAH_H
```

DLu replied:
I was going to argue with you about using modern standards, but then I dug and found [this bit of the Google guide](https://google.github.io/styleguide/cppguide.html#The__define_Guard). 

That's not going to stop me from posting a gif that came to mind. 

![I recognize the council has made a decision, but given that it's a stupid-ass decision, I've elected to ignore it.](https://i.pinimg.com/originals/54/8b/ca/548bca91b75d0911b697e8c2733996d0.gif)

(I'm not calling this a stupid decision, because upon further research, pragma is generally supported but is not official lore)

ahcorde commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,419 @@
+   pluginlib_export_plugin_description_file(rviz_common rviz_common_plugins.xml)
+
+
+* To generate the proper QT files, we need to
```
```suggestion
* To generate the proper Qt files, we need to
```

ahcorde commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,419 @@
+  private Q_SLOTS:
+    void updateStyle();
+
+* This gets called whenever the gui is changed, via QT's SLOT/SIGNAL framework.
```
```suggestion
* This gets called whenever the gui is changed, via Qt's SLOT/SIGNAL framework.
```

ahcorde commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,419 @@
+^^^^^^^^^^^
+
+
+* ``#include <rviz_common/properties/parse_color.hpp>`` - Contains helper function to convert property to ogre color.
```
```suggestion
* ``#include <rviz_common/properties/parse_color.hpp>`` - Contains helper function to convert property to OGRE color.
```

DLu committed:
Apply suggestions from code review

Co-authored-by: Alejandro Hernández Cordero <ahcorde@gmail.com>
Signed-off-by: David V. Lu!! <davidvlu@gmail.com>

DLu committed:
Include guards

DLu committed:
Merge branch 'rviz_plugin_tutorial' of github.com:DLu/ros2_documentation into rviz_plugin_tutorial

ahcorde reviewed (approved):
Thank you sir!

![200w](https://github.com/ros2/ros2_documentation/assets/1933907/99dde0b8-11e6-49d1-84e6-e16aa482db05)

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+
+* We're implementing the `MessageFilterDisplay <https://github.com/ros2/rviz/blob/0ef2b56373b98b5536f0f817c11dc2b5549f391d/rviz_common/include/rviz_common/message_filter_display.hpp#L43>`__ class which can be used with any message with a ``std_msgs/Header``.
+* The class is templated with our ``Point2D`` message type.
+* For reasons outside the scope of this tutorial, you need the ``Q_OBJECT`` macro in there to get the QT parts of the gui to work.
```
Could we possibly have a link to some Qt documentation that explains it?

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+* We're implementing the `MessageFilterDisplay <https://github.com/ros2/rviz/blob/0ef2b56373b98b5536f0f817c11dc2b5549f391d/rviz_common/include/rviz_common/message_filter_display.hpp#L43>`__ class which can be used with any message with a ``std_msgs/Header``.
+* The class is templated with our ``Point2D`` message type.
+* For reasons outside the scope of this tutorial, you need the ``Q_OBJECT`` macro in there to get the QT parts of the gui to work.
+* ``processMessage`` is the only method that needs to be implemented, which we'll do the cpp file.
```
```suggestion
* ``processMessage`` is the only method that needs to be implemented, which we'll do in the cpp file.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+
```
Unnecessary whitespace.

```suggestion
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+Building a Custom RViz Display
+==============================
+
+`rviz_plugin_tutorial <https://github.com/MetroRobots/rviz_plugin_tutorial>`__: A tutorial on how to make RViz plugins
```
I'll suggest we have a little more intro here, explaining what we are doing in this tutorial and why.  Similar to our other tutorials like http://docs.ros.org/en/rolling/Tutorials/Intermediate/Writing-an-Action-Server-Client/Cpp.html (the `Background` section in particular).

clalancette replied:
It would also be worthwhile mentioning here that we'll be changing branches on the above repository as we go along, since this is a different "structure" than most of the rest of the tutorials.

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+   PLUGINLIB_EXPORT_CLASS(rviz_plugin_tutorial::PointDisplay, rviz_common::Display)
+
+
+* The logging is not strictly necessary.
```
```suggestion
* The logging is not strictly necessary, but helps with debugging.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+CMakeLists.txt
+^^^^^^^^^^^^^^
+
+Everyone loves how verbose CMake files are in ROS 2! The following lines are on top of the standard boilerplate.
```
I'm not sure this belongs in the official tutorials like this.  Here's a rewrite:

```suggestion
Add the following lines to the top of the standard boilerplate.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+
+
+* To generate the proper Qt files, we need to
+  A) Turn ``CMAKE_AUTOMOC`` on
+  B) Wrap the headers by calling ``qt5_wrap_cpp`` with each header that has ``Q_OBJECT`` in it.
+  C) Include the ``MOC_FILES`` in the library alongside our other cpp files.
```
I think you can do this with `.rst` with `#.`.

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+  A) Turn ``CMAKE_AUTOMOC`` on
+  B) Wrap the headers by calling ``qt5_wrap_cpp`` with each header that has ``Q_OBJECT`` in it.
+  C) Include the ``MOC_FILES`` in the library alongside our other cpp files.
+* Note that if you do NOT wrap your header files, you may get an error message when attempting to load the plugin at runtime, along the lines of:
+  .. code-block::
```
```suggestion
* Note that if you do NOT wrap your header files, you may get an error message when attempting to load the plugin at runtime, along the lines of:

  .. code-block::
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+Testing it out
+^^^^^^^^^^^^^^
+
+Compile your code and run ``rviz2``. You should be able to add your new plugin by clicking ``Add`` in the bottom left, and then selecting your package/plugin.
```
```suggestion
Compile your code and run ``rviz2``.
You should be able to add your new plugin by clicking ``Add`` in the bottom left, and then selecting your package/plugin.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+   :alt: screenshot of error state
+
+
+Let's put the topic ``/point`` and it should load fine, although not display anything.
```
```suggestion
If we put the topic ``/point`` in, it should load fine but not display anything.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+
+
+* We need to get the proper frame for our message and transform the ``scene_node_`` accordingly. This ensures that the visualization does not always appear relative to the fixed frame.
+* The actual visualization that we've been building to this entire time is in the last four lines: we set the position of the visualization to match the message's position.
```
```suggestion
* The actual visualization that we've been building up to is in the last four lines: we set the position of the visualization to match the message's position.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+   :alt: screenshot of functioning display
+
+
+If the box does not appear in that location, it might be because
+
```
```suggestion
If the box does not appear in that location, it might be because:
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+
+You can view the full version of this step with the branch name ``step4``.
+
+You can also set the status of the display. As an arbitrary example, let's make our display show a warning when the x coordinate is negative, because why not? In ``processMessage``\ :
```
```suggestion
You can also set the status of the display.
As an arbitrary example, let's make our display show a warning when the x coordinate is negative, because why not?
In ``processMessage``:
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+
+     Ogre::Vector3 position;
+     Ogre::Quaternion orientation;
+     if (!context_->getFrameManager()->getTransform(msg->header, position, orientation))
+     {
```
This isn't our usual style for ROS 2.  The usual style for conditionals and loops is:

```suggestion
     if (!context_->getFrameManager()->getTransform(msg->header, position, orientation)) {
```

clalancette replied:
(same goes for a bunch of the other C++ code in this tutorial)

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+Cleanup
+-------
+
+Now its time to clean it up a bit. This makes things look nicer and be a little easier to use, but aren't strictly required. You can view the full version of this step with the branch name ``step5``.
```
```suggestion
Now its time to clean it up a bit.
This makes things look nicer and be a little easier to use, but aren't strictly required.
You can view the full version of this step with the branch name ``step5``.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,421 @@
+* We put actual text into the description. Don't be lazy.
+* By declaring the specific message type here, when you attempt to add a Display by Topic, it will suggest this plugin for the topics of that type.
+
+We also add an icon for the plugin at ``icons/classes/Point2D.png``. The folder is hardcoded, and the filename should match the name from the plugin declaration (or the name of the class if not specified). `[icon source] <https://commons.wikimedia.org/wiki/File:Free_software_icon.svg>`_
```
```suggestion
We also add an icon for the plugin at ``icons/classes/Point2D.png``.
The folder is hardcoded, and the filename should match the name from the plugin declaration (or the name of the class if not specified). `[icon source] <https://commons.wikimedia.org/wiki/File:Free_software_icon.svg>`_
```

clalancette reviewed (changes requested):
Overall, this is great.  Thanks for moving it around.

I left a bunch of stuff to cleanup.

DLu committed:
Apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Signed-off-by: David V. Lu!! <davidvlu@gmail.com>

DLu committed:
PR Feedback

DLu committed:
Whitespace

DLu commented:
I will update the code style if you can provide a `.clang-format` file that is to be used canonically for tutorials. 

I find this process frustrating, to have a PR sit untouched for a year, then get held up in the end for the placement of curly braces. This seems like perfect is the enemy of the good. If you want to do something to increase the number of community generated tutorials, you need to lower the bar **a little**.

clalancette commented:
> I will update the code style if you can provide a `.clang-format` file that is to be used canonically for tutorials.

Unfortunately, that is impossible; `clang-format` can't be configured to follow our coding style currently.

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,442 @@
+With the first option, there is more network traffic and limitations to how the data can be represented. It is also quick and flexible.
+The latter option is explained in this tutorial. It takes a bit of work, but can lead to much richer visualizations.
+
+All of the code for this tutorial can be found in `this repository <https://github.com/MetroRobots/rviz_plugin_tutorial>`__.
```
I couldn't find https://github.com/MetroRobots/rviz_plugin_tutorial ; is it a private repository or something?

DLu replied:
Public now, sorry.

clalancette commented:
> Unfortunately, that is impossible; `clang-format` can't be configured to follow our coding style currently.

Sorry, that was a little terse.  We can't do `clang-format`, but the right solution here is to enable `uncrustify` on the end repository, and then fix all of the uncrustify problems.  That will then conform to the style that we want.  I'm willing to do that work if the `rviz_plugin_tutorial` repository is available somewhere (that's a blocking issue anyway).

clalancette commented:
> Sorry, that was a little terse. We can't do `clang-format`, but the right solution here is to enable `uncrustify` on the end repository, and then fix all of the uncrustify problems. That will then conform to the style that we want. I'm willing to do that work if the `rviz_plugin_tutorial` repository is available somewhere (that's a blocking issue anyway).

All right, see https://github.com/MetroRobots/rviz_plugin_tutorial/pull/1

DLu committed:
Update code style

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+Boilerplate for Basic Plugin
+----------------------------
+
+Strap in, there's a lot of code. You can view the full version of this code with the branch name ``step1``.
```
```suggestion
Strap in, there's a lot of code.
You can view the full version of this code with the branch name ``step1``.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+
+* We're implementing the `MessageFilterDisplay <https://github.com/ros2/rviz/blob/0ef2b56373b98b5536f0f817c11dc2b5549f391d/rviz_common/include/rviz_common/message_filter_display.hpp#L43>`__ class which can be used with any message with a ``std_msgs/Header``.
+* The class is templated with our ``Point2D`` message type.
+* `For reasons outside the scope of this tutorial <https://doc.qt.io/archives/qt-4.8/moc.html>`__, you need the ``Q_OBJECT`` macro in there to get the QT parts of the gui to work.
```
Since we depend on Qt-5, let's point to the Qt5 version of this document:

```suggestion
* `For reasons outside the scope of this tutorial <https://doc.qt.io/qt-5/moc.html>`__, you need the ``Q_OBJECT`` macro in there to get the QT parts of the GUI to work.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+
+
+* The logging is not strictly necessary, but helps with debugging.
+* In order for RViz to find our plugin, we need this ``PLUGINLIB`` invocation in our code (as well as other things below)
```
```suggestion
* In order for RViz to find our plugin, we need this ``PLUGINLIB`` invocation in our code (as well as other things below).
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+
+* To generate the proper Qt files, we need to
+
+  * Turn ``CMAKE_AUTOMOC`` on
```
```suggestion
  * Turn ``CMAKE_AUTOMOC`` on.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+
+     [rviz2]: PluginlibFactory: The plugin for class 'rviz_plugin_tutorial::PointDisplay' failed to load. Error: Failed to load library /home/ros/ros2_ws/install/rviz_plugin_tutorial/lib/libpoint_display.so. Make sure that you are calling the PLUGINLIB_EXPORT_CLASS macro in the library code, and that names are consistent between this macro and your XML. Error string: Could not load library LoadLi ...
+
+* A lot of the other code ensures that the plugin portion works. Namely, calling ``pluginlib_export_plugin_description_file`` is essential to getting RViz to find your new plugin.
```
```suggestion
* A lot of the other code ensures that the plugin portion works.
Namely, calling ``pluginlib_export_plugin_description_file`` is essential to getting RViz to find your new plugin.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+   }
+
+
+* We need to get the proper frame for our message and transform the ``scene_node_`` accordingly. This ensures that the visualization does not always appear relative to the fixed frame.
```
```suggestion
* We need to get the proper frame for our message and transform the ``scene_node_`` accordingly.
This ensures that the visualization does not always appear relative to the fixed frame.
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+^^^^^^^^^^^^^^
+
+
+* ``#include <rviz_common/properties/color_property.hpp>`` Color is but one of many properties you can set.
+
+.. code-block:: c++
+
+  private Q_SLOTS:
+    void updateStyle();
+
+* This gets called whenever the gui is changed, via Qt's SLOT/SIGNAL framework.
+* ``std::unique_ptr<rviz_common::properties::ColorProperty> color_property_;`` The property itself.
```
I find the layout of this section a bit confusing.  I think we should do something more like:

```suggestion

Include the header file for color properties: ``#include <rviz_common/properties/color_property.hpp>``.
Color is but one of many properties you can set.

Add in the prototype for ``updateStyle``, which is called whenever the GUI is changed via Qt's SIGNAL/SLOT framework:

.. code-block:: c++

  private Q_SLOTS:
    void updateStyle();

Add in a new property to store the property itself: ``std::unique_ptr<rviz_common::properties::ColorProperty> color_property_;``
```

clalancette commented on source/Tutorials/Intermediate/RViz/RViz-Custom-Display/RViz-Custom-Display.rst:
```diff
@@ -0,0 +1,440 @@
+   </library>
+
+
+* We add the ``name`` field to the ``class`` tag. This changes the name that is displayed in RViz. In code, it makes sense to call it a ``PointDisplay`` but in RViz, we want to simplify.
```
```suggestion
* We add the ``name`` field to the ``class`` tag.
This changes the name that is displayed in RViz. In code, it makes sense to call it a ``PointDisplay`` but in RViz, we want to simplify.
```

DLu replied:
```suggestion
* We add the ``name`` field to the ``class`` tag.
This changes the name that is displayed in RViz.
In code, it makes sense to call it a ``PointDisplay`` but in RViz, we want to simplify.
```

clalancette reviewed (changes requested):
Thanks for sticking with this one.  I've left a few more grammatical/layout things to change, but this is looking close.

DLu committed:
Apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Signed-off-by: David V. Lu!! <davidvlu@gmail.com>

DLu committed:
Warning cleanup

clalancette reviewed (approved):
This is great.  Thanks @DLu for sticking with this.  I'm going to go ahead and merge this in, and also backport to `iron` and `humble`.

clalancette commented:
@Mergifyio backport humble iron
